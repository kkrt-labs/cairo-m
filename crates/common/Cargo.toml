[package]
name = "cairo-m-common"
version = "0.1.0"
edition.workspace = true

[dependencies]
num-traits.workspace = true
paste = "1.0"
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
smallvec.workspace = true

# Using workspace dependency for stwo
stwo-prover.workspace = true
