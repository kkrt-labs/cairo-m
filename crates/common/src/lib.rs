pub mod instruction;
pub mod opcode;
pub mod program;

pub use instruction::Instruction;
pub use opcode::Opcode;
pub use program::{Program, ProgramMetadata};
