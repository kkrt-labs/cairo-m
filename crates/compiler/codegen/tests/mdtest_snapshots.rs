//! Codegen snapshot tests for mdtest cases.
//! This file automatically generates codegen snapshots for all Cairo-M code in the mdtest directory,
//! providing comprehensive coverage of real-world examples through the entire compilation pipeline.

mod common;

use cairo_m_compiler_codegen::CodeGenerator;
use cairo_m_compiler_mir::generate_mir;
use cairo_m_test_utils::{mdtest::MdTestRunner, mdtest_path};
use common::{create_test_crate, TestDatabase};

#[test]
fn test_mdtest_codegen_snapshots() {
    use insta::{assert_snapshot, glob, with_settings};

    // Use glob! to discover all markdown files in mdtest directory
    glob!(mdtest_path().to_str().unwrap(), "**/*.md", |path| {
        let db = TestDatabase::default();

        let runner = MdTestRunner::new("CASM", |source, name| {
            let crate_id = create_test_crate(&db, source, name, "mdtest");

            // First generate MIR
            let mir_module = match generate_mir(&db, crate_id) {
                Ok(module) => module,
                Err(diagnostics) => {
                    return Err(format!("MIR generation failed: {:#?}", diagnostics));
                }
            };

            // Then generate CASM code from MIR
            let mut generator = CodeGenerator::new();
            match generator.generate_module(&mir_module) {
                Ok(_) => Ok(generator.debug_instructions()),
                Err(e) => Err(format!("Code generation failed: {:#?}", e)),
            }
        })
        .with_parent_dir(true);

        let snapshots = runner.run_file(path);

        for snapshot in snapshots {
            with_settings!({
                description => format!("Codegen snapshot for mdtest: {}", snapshot.name).as_str(),
                omit_expression => true,
                snapshot_suffix => snapshot.suffix,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(snapshot.content);
            });
        }
    });
}
