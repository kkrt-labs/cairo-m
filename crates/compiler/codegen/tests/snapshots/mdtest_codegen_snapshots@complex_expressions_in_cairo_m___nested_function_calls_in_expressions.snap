---
source: crates/compiler/codegen/tests/mdtest_snapshots.rs
description: "Codegen snapshot for mdtest: Complex Expressions in Cairo-M - Nested Function Calls in Expressions"
input_file: mdtest/01-basics/06-expressions.md
---
Source:
fn compute_base(x: felt) -> felt {
    return x * 2;
}

fn compute_offset(y: felt) -> felt {
    return y + 5;
}

fn test_function_expressions() -> felt {
    let a = 7;
    let b = 3;

    // Function calls in expressions
    let result1 = compute_base(a) + compute_offset(b);     // (7 * 2) + (3 + 5) = 14 + 8 = 22
    let result2 = compute_base(a + b) - compute_offset(1); // (10 * 2) - (1 + 5) = 20 - 6 = 14

    // Nested function calls
    let result3 = compute_base(compute_offset(a));         // compute_base(7 + 5) = compute_base(12) = 24

    return result1 + result2 + result3;                    // 22 + 14 + 24 = 60
}
============================================================
Generated CASM:
compute_base:
compute_base:
compute_base_0:
   0: 6 -4 2 -3            // [fp + -3] = [fp + -4] op 2
   1: 11 _ _ _             // return
compute_offset:
compute_offset:
compute_offset_0:
   2: 4 -4 5 -3            // [fp + -3] = [fp + -4] op 5
   3: 11 _ _ _             // return
test_function_expressions:
test_function_expressions:
test_function_expressions_0:
   4: 9 7 0 _              // [fp + 0] = 7
   5: 9 3 1 _              // [fp + 1] = 3
   6: 4 0 0 13             // Arg 0 slot 0: [fp + 13] = [fp + 0] + 0
   7: 10 15 0 _            // call compute_base
   8: 4 1 0 16             // Arg 0 slot 0: [fp + 16] = [fp + 1] + 0
   9: 10 18 2 _            // call compute_offset
  10: 0 14 17 4            // [fp + 4] = [fp + 14] op [fp + 17]
  11: 0 0 1 19             // [fp + 19] = [fp + 0] op [fp + 1]
  12: 10 21 0 _            // call compute_base
  13: 9 1 22 _             // Arg 0: [fp + 22] = 1
  14: 10 24 2 _            // call compute_offset
  15: 1 20 23 8            // [fp + 8] = [fp + 20] op [fp + 23]
  16: 4 0 0 25             // Arg 0 slot 0: [fp + 25] = [fp + 0] + 0
  17: 10 27 2 _            // call compute_offset
  18: 10 28 0 _            // call compute_base
  19: 0 4 8 11             // [fp + 11] = [fp + 4] op [fp + 8]
  20: 0 11 27 -3           // [fp + -3] = [fp + 11] op [fp + 27]
  21: 11 _ _ _             // return
